Initial commit of persistence extensions

cdjackson opened the pull request:
This contains initial thoughts for persistence extensions.

Two extensions are considered -:

1. An interface to write an HistoricState. This allows bindings to write data with a timestamp that is not the current time. This is needed to support devices that periodically provide a dump of historic data.
2. An admin interface that allows the database to be administered. This supports methods to return the current list of items that are persisted, and supports deleting items from the database and deleting data from an item.

I'm sure there are other ideas that can improve this, but hopefully this will help to prompt some discussion...

Signed-off-by: Chris Jackson <chris@cd-jackson.com>

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/HistoricalPersistenceStoreService.java:
```diff
@@ -0,0 +1,32 @@
+ * @author Chris Jackson - Initial implementation and API
+ *
+ */
+public interface HistoricalPersistenceStoreService {
```
The current interfaces are called PersistenceService and QueryablePersistenceService.
So "Store" is definitely too much here. I also wonder, if there is a better description than "Historical". How about ModifiablePersistenceService?

Should this interface directly extend QueryablePersistenceService or can there be real cases where it isn't queryable?

The PersistenceAdminService also wants to address removal of entries. If I remember our past discussions right, you probably want to add some delete/remove methods to this interface as well, don't you?

cdjackson replied:
> So "Store" is definitely too much here.

Ok - ```HistoricalPersistenceService```?

> How about ModifiablePersistenceService?

But it's not modifyable - it just allows you to store information with a time - you can't update anything in the store so in this respect it's no different than the other services.

> Should this interface directly extend QueryablePersistenceService

Yes - you're probably right - I don't think it makes sense without query.

> The PersistenceAdminService also wants to address removal of entries. If I remember our past discussions right, you probably want to add some delete/remove methods to this interface as well, don't you?

Yes - I split those into a separate interface though. I'm happy to consolidate into a single interface if you prefer, but I recall in the past you wanted to keep admin (delete) functions separate.

kaikreuzer replied:
> But it's not modifyable

Right, but if we combine it with your other "separate interface to come", it will be :-)

> I'm happy to consolidate into a single interface if you prefer, but I recall in the past you wanted to keep admin (delete) functions separate.

Well, I wanted to keep any non-chonological write operations separate. If we introduce a "store with timestamp" method, I think this anyhow means that the service is able to modify its already persisted data - and hence also offers deletion of entries etc. So for me, it makes sense to combine those methods in one interface.

cdjackson replied:
> Right, but if we combine it with your other "separate interface to come", it will be :-)

Yes, if we combine into a single interface, then this is fine with me.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/PersistenceAdminService.java:
```diff
@@ -0,0 +1,43 @@
+ * @author Chris Jackson - Initial implementation and API
+ *
+ */
+public interface PersistenceAdminService {
```
I assume you would want to register an implementation of this interface as an OSGi service and also built a REST endpoint for it, right?

kaikreuzer replied:
Do we at all need such a service or couldn't the PersistenceResource be enhanced to simply take ModifiablePersistenceServices into account and offer modification/deletion operations for them?

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/PersistenceAdminService.java:
```diff
@@ -0,0 +1,43 @@
+ */
+public interface PersistenceAdminService {
+    /**
+     * Returns a list of items that are stored in the persistence service
```
There is not "the persistence service", so what shall this method do?

cdjackson replied:
I'm not sure I understand what you mean? This is intended to be an interface that can be implemented by a persistence service and this method is meant to return a list of items that this persistence service has persisted.

kaikreuzer replied:
Ok, in this case the interface name was very misleading. It should simply also be a XXXPersistenceService, shouldn't it? So as discussed above -> combine those into a single interface.

cdjackson commented:
Ok - so I will update this PR with a single ```ModifyablePersistenceService``` interface as above. I'll also look to extend the REST interface to support the extra methods.

I'll also look to see if I can port a single persistence service to ESH to implement these services. This might take a little thought - ideally I would have done the JDBC, but there might be license issues and I might need to break out the drivers into separate repos to make it ESH compatible.

kaikreuzer commented:
> with a single ModifyablePersistenceService interface as above

You are the native speaker, but isn't the correct spelling Modifiable?

> I'll also look to see if I can port a single persistence service to ESH to implement these services.

What about your H2 efforts you were once doing? As it is available under EPL, this sounds like a pretty good candidate?

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/PersistenceAdminService.java:
```diff
@@ -0,0 +1,43 @@
+     * Returns a list of items that are stored in the persistence service
+     *
+     * This is returned as a string to allow the persistence service to return items that are no long available as an
+     * openHAB {@link Item}.
```
please don't mention openHAB here

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/PersistenceAdminService.java:
```diff
@@ -0,0 +1,43 @@
+     * This is returned as a string to allow the persistence service to return items that are no long available as an
+     * openHAB {@link Item}.
+     *
+     * @return list of strings of item names contained in the store
```
we usually return Sets (not Lists or Iterables)

kaikreuzer replied:
If we really require this method, shouldn't it go to QueryablePersistenceService?

cdjackson replied:
I think it's useful to allow maintenance of a database. I'm sure we've all created items that have had data persisted, and we've then deleted the item and created another with a different name. I think we need some way to list what is in there without reference to the item registery so it can be removed.

I'm happy to add this to the QueryablePersistenceService - my initial thought was to avoid changing the current interfaces but if you're happy that this is consolidated with the move to ESH, then I'll do that.

kaikreuzer replied:
The good thing is that I can deal with the new method within the compat layer, so changing the interface does not mean that the existing 1.x services would break. So yes, let's put the methods there where it makes most sense.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/PersistenceAdminService.java:
```diff
@@ -0,0 +1,43 @@
+     *
+     * @return true if the item was deleted
+     */
+    public boolean removeItem(String itemName);
```
I'd call it something like "removeAllData(String itemName)"

cdjackson commented:
> You are the native speaker, but isn't the correct spelling Modifiable?

Native speaker, yes, but also an engineer, so be definition, spelling is far from my strong point! :)

cdjackson commented:
API updated -:

* Changes made to ```QueryablePersistenceService``` and ```ModifiablePersistenceService``` as discussed
* ```store``` method changed to add a time as a parameter rather than to use HistoricalItem. This didn't work well since it didn't include the Item - only the item name, and that wasn't ideal as the item type is needed in the persistence service.
* ```getItems``` method updated to return a custom type ```PersistenceItem``` which allows more than just the name to be returned. Currently I have added the item name, and number of rows.

Any thoughts on this appreciated. I've already made most of the mods to the persistence service and added the H2SQL service - more testing and tidying required of course, but before I do I thought I'd get feedback on the API.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/ModifiablePersistenceService.java:
```diff
@@ -0,0 +1,52 @@
+     *
+     * @return true if the item was deleted
+     */
+    public boolean removeItemData(String itemName);
```
to make it very clear, I would call it `removeAllItemData`.

cdjackson replied:
I am considering to remove this method and instead just have the ```remove``` method. If all data is removed using the ```remove``` method, then the item should also be deleted from the database.  I think this is a more fail safe method - it means you need to actively configure a filter with a start/stop time that includes all data.

What do you think?

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/PersistenceItem.java:
```diff
@@ -0,0 +1,14 @@
+ */
+package org.eclipse.smarthome.core.persistence;
+
+public interface PersistenceItem {
```
I would prefer some other name to better express the purpose/content.
How about something like PersistedItemInfo or so?

kaikreuzer commented:
lgtm, so fine for me if you go ahead. @maggu2810 Any comments from your end?

maggu2810 commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/ModifiablePersistenceService.java:
```diff
@@ -0,0 +1,52 @@
+     * @param date the date of the record
+     * @param item the data to be stored
+     */
+    public void store(Date date, Item item);
```
If you don't care, I would prefer to use Item as first argument and Date as second one.
Why? Because there is already the method `void store(Item item);`.
https://github.com/eclipse/smarthome/blob/e0985e9/bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/PersistenceService.java#L40
And if this one has an additional argument I prefer to add it at the endt.

maggu2810 replied:
We can remove "public" modifiers on interface functions (its default).

cdjackson replied:
I know - the reason I added them was to maintain the same style as other interfaces in this package. I’m happy to remove it though.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/PersistenceItem.java:
```diff
@@ -0,0 +1,14 @@
+public interface PersistenceItem {
+    String getName();
+
+    Integer getRows();
```
Why "rows"? Shouldn't we use "count" or "size"?

cdjackson replied:
We could - I don’t mind. rows is common database terminology which is why I used it.

maggu2810 replied:
You are right, that rows is used by SQL databases. IIRC it differs using NOSQL databases.
The interface is about persistence, not about SQL databases.
If rows or files or map entries are used is implementation dependent.

cdjackson replied:
So? It needs to be called something - it wasn't planned to be SQL dependant - other databases use this term as well.
As I said - I don't care, but I don't see anything wrong with rows...

cdjackson commented:
Updated following latest comments. As per previous comment, I've now removed the method to delete the item. It is expected that the persistence service should free resources associated with an item if all data is deleted.

I've also added the REST changes if you have comments here.

H2SQL implementation to follow.

cdjackson commented:
So, there is another issue with the ```store(item, time)``` method. The problem is that you can't use it as you can't set the state of an item, and even if you could, you wouldn't want to set it to an old value. In order to use this method, the implementer would need to create a class that implements ```Item```, and this seems like a lot of unnecessary hassle.

So, options...

1) Revert to my original idea of using the ```store(HistoricState)``` method. This has the problem that persistence services probably want to know the item type (all the SQL related services do at least), and that's not available in ```HistoricState```.
2) Implement ```store(item, HistoricState)```.
3) Implement ```store(item, time, state)```.
4) Your idea here...

maggu2810 commented:
I assume 2) or 3) should be fine.

maggu2810 commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/PersistenceItemInfo.java:
```diff
@@ -0,0 +1,36 @@
+     *
+     * @return count of the number of rows of data.
+     */
+    Integer getRows();
```
See: https://github.com/eclipse/smarthome/pull/1872/files/b532fa51461caa8855eedacb3e203fad9b9d5e59#r71975733

maggu2810 commented on bundles/io/org.eclipse.smarthome.io.rest.core/META-INF/MANIFEST.MF:
```diff
@@ -40,6 +40,7 @@ Import-Package: com.google.common.base,
  org.eclipse.smarthome.io.rest,
  org.eclipse.smarthome.io.rest.core.item,
  org.eclipse.smarthome.io.rest.core.thing,
+ org.openhab.core.items,
```
ESH should not rely on an OH package

maggu2810 commented on extensions/persistence/org.eclipse.smarthome.persistence.h2sql/.classpath:
```diff
@@ -0,0 +1,8 @@
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.6"/>
+	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
+	<classpathentry kind="src" path="java"/>
+	<classpathentry kind="lib" path="lib/h2-1.3.176.jar"/>
```
There are CQs for other Eclipse projects that are using h2.
The most recent CQ is for "1.4.186".
If we need to use a 1.3 one, the most recent here is: 1.3.169

I assume the CQ will be applied more recently, if we use a version another approved CQ already exist.

cdjackson replied:
We can, but I am using the most recent stable version. We can use a beta version if you wish, and I don't mind using 1.4.186, but thought that for such a product that may be used commercially, that using beta libraries might be frowned upon?

maggu2810 commented on extensions/persistence/org.eclipse.smarthome.persistence.h2sql/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,24 @@
+ org.eclipse.smarthome.core.library.types,
+ org.eclipse.smarthome.core.persistence,
+ org.eclipse.smarthome.core.types,
+ org.openhab.core.library.types,
```
Same here, we cannot rely on an OH package

cdjackson replied:
No problem - it's just an error as I originally wrote the H2 persistence forn the OH2 repo.

cdjackson commented:
Ok - updated and ready for review.

As a summary of the REST API -:
![image](https://cloud.githubusercontent.com/assets/1041475/17084564/d3f9671a-51b8-11e6-9ec3-27140828e4e9.png)

I've still got the current stable of H2SQL included - if you would prefer to use a more recent (beta) then let me know and I can change.

maggu2810 commented:
> I've still got the current stable of H2SQL included - if you would prefer to use a more recent (beta) then let me know and I can change.

Sorry about that, I did not check which version is the most recent stable and which is the beta one.
I only looked at the CQs that are already present.

I am fine with the one (most recent stable) that you have chosen.

cdjackson commented:
I've just changed the way the GET works when doing a query. Previously, this was designed specifically for charting where you really need to get the value just before, and just after the requested period, as well as the requested data. The reason for this is if you are only logging changes, and there are no changes in the charted period, then there is no data returned.

However, while this works well for charts, it's less desirable when scrolling through the data in a tabular form. So, I've added a parameter to control this. Setting boundary=true will return the value before and after the requested period and is best suited for charting. (maybe the name may be better - suggestions welcome!).

One point to note - the before and after values will be outside of the period requested. I did it this way as it would allow chart implementations to interpolate the data (if desired) to the start time. However, it is different to the current implementation which is set exactly to the requested start and stop times. Both have advantages and disadvantages, so I welcome comments on preferences.

I've implemented an editor in HABmin to view the data in a table, delete, and update entries. This is using the H2 implementation in this PR and it seems to work pretty well.

kaikreuzer commented:
Might I suggest to split the h2 persistence service implementation into a separate PR? We could then as a first step merge this PR here (with the content it initially describes), while we will need a bit more time for the h2 PR, since it will require 1 or 2 CQs to be approved.

cdjackson commented:
Done - I'll add a separate PR for H2.

kaikreuzer commented:
Thanks, that was quick. Will do a final review soon (didn't yet look at the REST parts)

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/ModifiablePersistenceService.java:
```diff
@@ -0,0 +1,53 @@
+/**
```
could you re-run "mvn license:format" to update the headers to the latest version?

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/QueryablePersistenceService.java:
```diff
@@ -7,20 +7,34 @@
+     *
+     * @return list of strings of item names contained in the store. Not null.
+     */
+    Set<PersistenceItemInfo> getItems();
```
Better rename the method to `getItemInfos()`.

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/persistence/PersistenceItemInfoBean.java:
```diff
@@ -0,0 +1,39 @@
+ * @author Chris Jackson - Initial Contribution
+ *
+ */
+public class PersistenceItemInfoBean {
```
Note that we by now always call such beans DTOs and where ever they might be relevant for serialization anywhere else than the REST API as well (e.g. for the payload of events), we put those DTOs in the core bundle in a sub-package "dto" (here it would be org.eclipse.smarthome.core.persistence.dto).

Could you possibly refactor the existing "Beans" in this package to follow this convention?
Regarding the PersistenceItemInfoBean: Do we need this at all? It is very similar to the PersistenceItemInfo class - maybe you could add a simply GSON annotation to the Date fields to specify the string format in which they should be serialized?

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/persistence/PersistenceResource.java:
```diff
@@ -58,18 +68,22 @@
  * store
  *
  * @author Chris Jackson - Initial Contribution
+ * @author Chris Jackson - Updated to add support for ModifiablePersistenceService
```
Don't add a second line with the same author - rather combine the description text (although initial contrib is still good enough as long as you are the only author).

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/persistence/PersistenceResource.java:
```diff
@@ -58,18 +68,22 @@
 
     private final Logger logger = LoggerFactory.getLogger(PersistenceResource.class);
     private final int MILLISECONDS_PER_DAY = 86400000;
+    private final SimpleDateFormat isoDateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ");
```
You could also refer to 'DateTimeType.DATE_PATTERN_WITH_TZ_AND_MS' if we want to make sure that we use a similar format in both cases.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/ModifiablePersistenceService.java:
```diff
@@ -0,0 +1,53 @@
+
+    /**
+     * Removes data associated with an item from a persistence service.
+     * If all data is removed for the specified item, the persistence service should free any resources associated with
```
Please update the javadoc - the filtercriteria might not filter for a specific item at all, so this method is applicable for many items at the same time as well.

cdjackson replied:
I'm not sure I follow. I assumed that there must be an item name provided - it's not possible to add a filter with multiple item names, and I think it would be dangerous to add a filter with no item name. As discussed elsewhere, in the REST interface at least, it is required. Technically it would be possible to implement a filter that removed data from multiple items, but is it a good idea?

kaikreuzer replied:
Well, if you look at FilterCriteria, it reads:
"...properties are constraints on the query result. These properties include the item name, begin and end date and the item state" and "filter result to only contain entries for the given item".

So in general, it allows that the itemName is null and hence the filter is applicable for ANY item. As your method accepts a filter criteria as a parameter, you have to deal with that fact. If you do not want to allow itemName to be null, you should document this in the JavaDoc and declare an exception that you will throw in that case.

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/persistence/PersistenceResource.java:
```diff
@@ -95,14 +118,57 @@ public Response httpGetPersistenceServices(@Context HttpHeaders headers) {
+            @ApiResponse(code = 404, message = "Unknown persistence service") })
+    public Response httpDeletePersistenceServiceItem(@Context HttpHeaders headers,
+            @ApiParam(value = "Name of the persistence service.", required = true) @QueryParam("servicename") String serviceName,
+            @ApiParam(value = "The item name.", required = true) @PathParam("itemname") String itemName,
```
Might we also need to have a url without having to provide an item name? For functionality like "delete all data older than 12 months"? Or is this too risky to expose through REST or not a usecase for now?

cdjackson replied:
It is certainly a use case, but I considered it too risky. My philosophy for deleting information was to make the user provide as much information as possible - ie everything - start/finish times, item name and persistence service - no defaults - to make it as foolproof as possible.
I'd hate for someone to accidentally call the delete function, forget to add the item name, and delete all their data :)

kaikreuzer replied:
ok for me then

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/persistence/PersistenceResource.java:
```diff
@@ -95,14 +118,57 @@ public Response httpGetPersistenceServices(@Context HttpHeaders headers) {
+    public Response httpDeletePersistenceServiceItem(@Context HttpHeaders headers,
+            @ApiParam(value = "Name of the persistence service.", required = true) @QueryParam("servicename") String serviceName,
+            @ApiParam(value = "The item name.", required = true) @PathParam("itemname") String itemName,
+            @ApiParam(value = "Start time of the data to return.", required = true) @QueryParam("starttime") String startTime,
```
as this param is merely a string, it would be helpful to mention in the description, which format is expected.

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/persistence/PersistenceResource.java:
```diff
@@ -124,12 +190,12 @@ private Response getItemHistoryBean(String serviceName, String itemName, String
         if (service == null) {
-            logger.debug("Persistence service not found '{}'.", serviceName);
+            logger.warn("Persistence service not found '{}'.", serviceName);
             return JSONResponse.createErrorResponse(Status.CONFLICT, "Persistence service not found: " + serviceName);
```
Shouldn't this be better a BAD_REQUEST response?

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/persistence/PersistenceResource.java:
```diff
@@ -124,12 +190,12 @@ private Response getItemHistoryBean(String serviceName, String itemName, String
         if (!(service instanceof QueryablePersistenceService)) {
-            logger.debug("Persistence service not queryable '{}'.", serviceName);
+            logger.warn("Persistence service not queryable '{}'.", serviceName);
             return JSONResponse.createErrorResponse(Status.CONFLICT,
```
Shouldn't this be better a BAD_REQUEST response?

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/persistence/PersistenceResource.java:
```diff
@@ -232,10 +313,160 @@ private Response getItemHistoryBean(String serviceName, String itemName, String
+        // If serviceName is null, then use the default service
+        PersistenceService service = null;
+        if (serviceName == null) {
+            // TODO: Add handler for default service once this is available in ESH
```
what is missing in ESH? Do you have an issue reference?

cdjackson replied:
I'm not sure if there is a reference. We had this discussion when I first added the persistence REST interface such that the user didn't need to request the service, and we would use the default. However there was no easy way to read the default service - I think this is only available in the persistence manager (from memory, which might be wrong).

We basically need a system wide method for reading the default persistence service.

kaikreuzer replied:
According to [this](http://stackoverflow.com/a/24628894), it should be possible to register as a ManagedService for the PID org.eclipse.smarthome.persistence and get the same configuration data as the PersistenceExtensions does. This would probably be the best solution, if it works.

cdjackson replied:
This doesn't seem to work. I added the managed service to the persistence resource - it gets called, but always with null parameter. I don't get the parameters with the default persistence being set until I remove the pid from the persistence model.

Any thoughts? We could add a ```PersistenceExtensions.getDefaultService()``` method, but that's another dependancy...

kaikreuzer replied:
> We could add a PersistenceExtensions.getDefaultService() method

For the moment, this is probably the simplest, if the theory above did not work out...
Another option would be to introduce a PersistenceConfigService which gets configured and the other services depend on it - but feels like too much overhead...

cdjackson replied:
Ok - I’ll update with this...

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/persistence/PersistenceResource.java:
```diff
@@ -232,10 +313,160 @@ private Response getItemHistoryBean(String serviceName, String itemName, String
+        }
+        if (service == null) {
+            logger.warn("Persistence service not found '{}'.", serviceName);
+            return JSONResponse.createErrorResponse(Status.CONFLICT, "Persistence service not found: " + serviceName);
```
BAD_REQUEST

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/persistence/PersistenceResource.java:
```diff
@@ -232,10 +313,160 @@ private Response getItemHistoryBean(String serviceName, String itemName, String
+
+        if (!(service instanceof QueryablePersistenceService)) {
+            logger.warn("Persistence service not queryable '{}'.", serviceName);
+            return JSONResponse.createErrorResponse(Status.CONFLICT,
```
BAD_REQUEST

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/persistence/PersistenceResource.java:
```diff
@@ -232,10 +313,160 @@ private Response getItemHistoryBean(String serviceName, String itemName, String
+        // For deleting, we must specify a servicename - don't use the default service
+        if (serviceName == null || serviceName.length() == 0) {
+            logger.warn("Persistence service must be specified for delete operations.");
+            return JSONResponse.createErrorResponse(Status.CONFLICT,
```
BAD_REQUEST

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/persistence/PersistenceResource.java:
```diff
@@ -232,10 +313,160 @@ private Response getItemHistoryBean(String serviceName, String itemName, String
+        PersistenceService service = persistenceServices.get(serviceName);
+        if (service == null) {
+            logger.warn("Persistence service not found '{}'.", serviceName);
+            return JSONResponse.createErrorResponse(Status.CONFLICT, "Persistence service not found: " + serviceName);
```
BAD_REQUEST

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/persistence/PersistenceResource.java:
```diff
@@ -232,10 +313,160 @@ private Response getItemHistoryBean(String serviceName, String itemName, String
+
+        if (!(service instanceof ModifiablePersistenceService)) {
+            logger.warn("Persistence service not modifiable '{}'.", serviceName);
+            return JSONResponse.createErrorResponse(Status.CONFLICT,
```
BAD_REQUEST

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/persistence/PersistenceResource.java:
```diff
@@ -124,12 +190,12 @@ private Response getItemHistoryBean(String serviceName, String itemName, String
 
         if (service == null) {
-            logger.debug("Persistence service not found '{}'.", serviceName);
+            logger.warn("Persistence service not found '{}'.", serviceName);
```
I would prefer to keep all logging to debug only - there is nothing wrong on the server side, when invalid requests come in, so we should not clutter the log. It should rather be the client who has to deal with the error situation (and possibly provide consolidated logs through https://github.com/eclipse/smarthome/pull/759).

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/persistence/ServiceBean.java:
```diff
@@ -17,5 +17,13 @@
+    /**
+     * Persistence service class
+     */
+    public String classname;
```
A classname is imho not appropriate to be exposed to the client side.
We should rather have a field "type", which could be "Standard", "Queryable" and "Modifiable" (or even only 0,1,2 if this is nothing that we ever would want to display directly to the user).

Btw, I think it would be a good time to "upgrade" PersistenceServices to ESH conventions wrt exposing them to UIs - we usually have "Ids" for services (which would be the current "name" field) and a label, which can be localized and returns the label that should be used in UIs (see e.g. [here](https://github.com/kaikreuzer/smarthome/blob/310c041a5309b28970948387bec66e2353ef0629/bundles/core/org.eclipse.smarthome.core.voice/src/main/java/org/eclipse/smarthome/core/voice/TTSService.java), where I have upgraded the TTSService interface in a similar way.

cdjackson replied:
Exposing a service name (ie label) is also something I was thinking about - if you're ok, I'll add a 'getLabel()' method to the PersistenceService class?

kaikreuzer replied:
Yes please - and possibly rename name to id if it does not have too much impact (which I don't think).

cdjackson replied:
No problem.

cdjackson commented:
@kaikreuzer all updates done I think. The only outstanding issue should be the default. Currently I've got the managed service code added, but as per previous comment, this doesn't appear to work so we need to solve that still....

cdjackson commented:
Ok - all done...

One issue I just realised is that there was an overlap with the REST api - if someone had an item called ```items``` it would have caused an issue, so I changed the API slightly. This is a breaking change from the previous persistence API, but given that was only merged a month or so back, and HABmin is probably the only software using it, I think it's best to change it.

Otherwise, this is running, and if you're happy I'll create an H2 PR as that's also ready to go.

cdjackson commented:
Looks like I missed a reference in bundles/ui/org.eclipse.smarthome.ui/src/main/java/org/eclipse/smarthome/ui/internal/chart/DefaultChartProvider.java. I'll update this tonight.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/ModifiablePersistenceService.java:
```diff
@@ -0,0 +1,55 @@
+     *
+     * @param filter the filter to apply to the data removal. ItemName can not be null.
+     * @return true if the query executed successfully
+     * @throws {@link InvalidParameterException} if item name is null.
```
Err, no: "This exception, designed for use by the JCA/JCE engine classes".
Better use IllegalArgumentException instead.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/dto/PersistenceItemInfoBean.java:
```diff
@@ -0,0 +1,39 @@
+ * @author Chris Jackson - Initial Contribution
+ *
+ */
+public class PersistenceItemInfoBean {
```
Call it DTO, not Bean. So did you come to the conclusion that you definitely need it?

cdjackson replied:
No - it shouldn't be used any more. I thought I'd deleted it but maybe I didn't commit it. Anyway, it's not needed. 

Sent from my iPhone

> On 2 Aug 2016, at 09:35, Kai Kreuzer <notifications@github.com> wrote:
> 
> In bundles/core/org.eclipse.smarthome.core.persistence/src/main/java/org/eclipse/smarthome/core/persistence/dto/PersistenceItemInfoBean.java:
> 
> > +/**
> > + * Copyright (c) 2014-2016 by the respective copyright holders.
> > + * All rights reserved. This program and the accompanying materials
> > + * are made available under the terms of the Eclipse Public License v1.0
> > + * which accompanies this distribution, and is available at
> > + * http://www.eclipse.org/legal/epl-v10.html
> > + */
> > +package org.eclipse.smarthome.core.persistence.dto;
> > +
> > +/**
> > + * This is a java bean that is used to serialize information about items in a persistence service to JSON.
> > + *
> > + * @author Chris Jackson - Initial Contribution
> > + *
> > + */
> > +public class PersistenceItemInfoBean {
> Call it DTO, not Bean. So did you come to the conclusion that you definitely need it?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/persistence/PersistenceResource.java:
```diff
@@ -69,68 +81,129 @@
 
-    private Response getItemHistoryBean(String serviceName, String itemName, String timeBegin, String timeEnd,
-            int pageNumber, int pageLength) {
+    private Response getItemHistoryDto(String serviceName, String itemName, String timeBegin, String timeEnd,
```
DTO not Dto

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/persistence/PersistenceResource.java:
```diff
@@ -201,41 +283,187 @@ private Response getItemHistoryBean(String serviceName, String itemName, String
+            PersistenceService persistence = service.getValue();
+            serviceDto.label = persistence.getLabel();
+            if (persistence instanceof ModifiablePersistenceService) {
+                serviceDto.type = "Modifiable";
```
Could you define them as constants at the top?

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/persistence/PersistenceResource.java:
```diff
@@ -201,41 +283,187 @@ private Response getItemHistoryBean(String serviceName, String itemName, String
         for (Map.Entry<String, PersistenceService> service : persistenceServices.entrySet()) {
-            ServiceBean serviceBean = new ServiceBean();
-            serviceBean.name = service.getKey();
+            PersistenceServiceDTO serviceDto = new PersistenceServiceDTO();
```
better call it serviceDTO

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/persistence/PersistenceResource.java:
```diff
@@ -201,41 +283,187 @@ private Response getItemHistoryBean(String serviceName, String itemName, String
 
-        return beanList;
+        mService.store(item, dateTime, state);
+        return JSONResponse.createResponse(Status.OK, "", "");
```
if you do not have a message to tell, you don't need to construct a JSONResponse. Use `Response.status(Status.OK).build();` instead.

kaikreuzer commented on bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/persistence/PersistenceResource.java:
```diff
@@ -69,68 +81,129 @@
+        this.itemRegistry = null;
     }
 
     @GET
```
As PersistenceServices now have a localizeable label, you should accept a locale as an optional parameter as well (see e.g. [here](https://github.com/eclipse/smarthome/blob/e0985e90d389350580e59bfbcd7e0adaaccf7519/bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/thing/ThingTypeResource.java#L110))

cdjackson commented:
Updated, and hopefully good to go ;)

kaikreuzer commented:
Thanks, looks all good to me!
The only one still complaining is Travis with a compilation error in TestPersistenceService...

cdjackson commented:
Sorry- missed that one... Fixed now ;)

kaikreuzer commented:
Wonderful 👍

kaikreuzer commented:
S%$#t, @cdjackson, I didn't notice that PersistenceExtensions is in smarthome.model, not in smarthome.core, so using it created an invalid dependency, see https://openhab.ci.cloudbees.com/job/openHAB-Core/60/console:
```
Unable to resolve org.eclipse.smarthome.io.rest.core/0.9.0.201608022010: missing requirement [org.eclipse.smarthome.io.rest.core/0.9.0.201608022010] osgi.wiring.package; filter:="(osgi.wiring.package=org.eclipse.smarthome.model.persistence.extensions)"
```
So we will need a different solution after all.
How about introducing a core.persistence.PersistenceServiceRegistry, where we can centrally offer methods like providing all services of a certain type, configuring the default, etc.? The other services could then refer to this and it might be useful in future anyhow (as not every service has to track all the different PersistenceServices itself).

cdjackson commented:
Other than for historical reasons, is there a reason that the persistence extensons are in the model bundle? IMHO, they should probably be part of the core.persistence bundle anyway.

I don't mind adding a registry, but I'm just wondering if we shouldn't do a slightly bigger refactor and move the PersistenceExtensions into the core bundle anyway, and then add the registry in with that? That would keep the model dealing with the model stuff only, and this is probably (??) a closer architecture to other models (like items at least) although I've not looked at them all.

I'm not sure if that would break anything (further!) or not - it just seems to be to be more logical and I suspect if we want to change it, now would be the time?

kaikreuzer commented:
If you check the history, you will actually notice that it [had been in core before](https://github.com/openhab/openhab/blob/v1.8.3/bundles/core/org.openhab.core.persistence/src/main/java/org/openhab/core/persistence/extensions/PersistenceExtensions.java) and was moved to model on purpose.
The reason was to keep all DSL related stuff in the model bundles and although the PersistenceExtensions is specifically there to provide methods for the DSL rules.
So no, moving it is not an option, especially not as long as it depends on jodatime, which should not be a core part either.

kaikreuzer commented:
Shall I work on this? It is rather critical as the OH distro build is currently broken by it...

cdjackson commented:
Ok - I'd assumed that these extensions were also going to be available in other rules as well - ie not just the 'old' DSL rules...

> Shall I work on this?

Sure - I can't do anything until later in the day, so if you have the time today then it might be safer (he says, knowing no-one really has 'time' :) ).

kaikreuzer commented:
>  I'd assumed that these extensions were also going to be available in other rules as well

Maybe only in some changed form and definitely without joda (hoping we then have Java8 instead) - so this refactoring might be done later.

> no-one really has 'time'

How true :-/ Anyhow, I'll look into it now!

kaikreuzer commented:
@cdjackson I have created https://github.com/eclipse/smarthome/pull/1962.
Might I ask you to review it?

cdjackson commented:
Sure - will be done early evening...
